runtime/compiler/optimizer/J9RecognizedCallTransformer.cpp

```diff
@@ -692,6 +694,33 @@ void J9::RecognizedCallTransformer::processUnsafeAtomicCall(TR::TreeTop* treetop
       if (enableTrace)
          traceMsg(comp(), "Created isObjectNull test node n%dn, non-null object will fall through to Block_%d\n", isObjectNullNode->getGlobalIndex(), treetop->getEnclosingBlock()->getNumber());
 
+      // Test if object is array - offheap only
+   #if defined (J9VM_GC_ENABLE_SPARSE_HEAP_ALLOCATION)
+      if (TR::Compiler->om.isOffHeapAllocationEnabled() && comp()->target().is64Bit())
+         {
+         //generate array check treetop
+         TR::Node *constNode = TR::Node::lconst(unsafeCall, ~(J9_SUN_FIELD_OFFSET_MASK));
+         TR::Node *unsafeAddress = TR::Node::create(TR::aladd, 2, objectNode, TR::Node::create(TR::land, 2, offsetNode, constNode));
```

Comment:
how can you do `aladd `before you know if it is an array or not? it is unclear to me what purpose this `aladd `is for.

Reply:
i seemed starting to understand what you are trying to do here.  if i understood it correctly, you should not try to do/load later from this `aladd` address, since the content in that memory slot is irrelevant to the call at all but the slot address is the target for the call.  for example:  compareAndSwap(objectNode, offsetNode, ...)  ...  the shape and form of content in [objectNode, offsetNode] doesn't need to be checked for the CAS operation, but objectNode itself matters for off-heap.  i.e.  you went down one more level.